query/marshal.go

```diff
@@ -150,6 +144,69 @@ func repoBranchesDecode(b []byte) (map[string][]string, error) {
 	return repoBranches, r.err
 }
 
+func repoBranchesIDsEncode(repoBranchesIDs map[string]*roaring.Bitmap) ([]byte, error) {
+	var b bytes.Buffer
+	var enc [binary.MaxVarintLen64]byte
+	varint := func(n uint64) {
+		m := binary.PutUvarint(enc[:], n)
+		b.Write(enc[:m])
+	}
+	str := func(s string) {
+		varint(uint64(len(s)))
+		b.WriteString(s)
+	}
+	strSize := func(s string) uint64 {
+		return uint64(binary.PutUvarint(enc[:], uint64(len(s))) + len(s))
+	}
+
+	// Calculate size
```

Comment:
How much does this precise size calculation matter? I can see it slightly reducing garbage, but with the other changes it should be pretty minimal, and it takes longer to do the encoding twice like this.

Reply:
It's likely not needed for IDs. For the older string based encoder it was very important due to the number of small objects.

Reply:
Seems worth it.

```
query:space-optimized-repo-sets* λ benchstat seqids.txt nosize.txt
name                       old time/op    new time/op    delta
RepoBranches_Encode-12        1.37s ±13%     1.31s ± 6%      ~     (p=0.548 n=5+5)
RepoBranches_Decode-12        1.27s ±17%     1.15s ± 2%    -9.71%  (p=0.016 n=5+4)
RepoBranchesIDs_Encode-12     168µs ± 9%     508µs ±12%  +201.76%  (p=0.008 n=5+5)
RepoBranchesIDs_Decode-12     301µs ±11%     289µs ± 2%      ~     (p=0.222 n=5+5)
RepoSet_Encode-12             1.27s ±20%     1.37s ±14%      ~     (p=0.151 n=5+5)
RepoSet_Decode-12             1.47s ±19%     1.58s ±12%      ~     (p=0.095 n=5+5)
RepoSetIDs_Encode-12         14.2µs ± 5%    13.6µs ± 5%      ~     (p=0.151 n=5+5)
RepoSetIDs_Decode-12         39.5µs ± 2%    40.3µs ± 2%    +2.13%  (p=0.016 n=5+5)

name                       old bytes      new bytes      delta
RepoBranches_Encode-12        189MB ± 0%     189MB ± 0%      ~     (all equal)
RepoBranchesIDs_Encode-12     443kB ± 0%     443kB ± 0%      ~     (all equal)
RepoSet_Encode-12             187MB ± 0%     187MB ± 0%      ~     (all equal)
RepoSetIDs_Encode-12         17.7kB ± 0%    17.7kB ± 0%      ~     (all equal)

name                       old alloc/op   new alloc/op   delta
RepoBranches_Encode-12        189MB ± 0%     189MB ± 0%      ~     (all equal)
RepoBranches_Decode-12        766MB ± 0%     766MB ± 0%      ~     (p=0.151 n=5+5)
RepoBranchesIDs_Encode-12     453kB ± 0%    1690kB ± 0%  +272.77%  (p=0.008 n=5+5)
RepoBranchesIDs_Decode-12     920kB ± 0%     920kB ± 0%      ~     (p=0.167 n=5+5)
RepoSet_Encode-12             226MB ± 0%     226MB ± 0%      ~     (p=0.333 n=5+4)
RepoSet_Decode-12             536MB ± 0%     536MB ± 0%      ~     (p=0.548 n=5+5)
RepoSetIDs_Encode-12         40.5kB ± 0%    40.5kB ± 0%      ~     (all equal)
RepoSetIDs_Decode-12         47.1kB ± 0%    47.1kB ± 0%      ~     (all equal)

name                       old allocs/op  new allocs/op  delta
RepoBranches_Encode-12         1.00 ± 0%      1.00 ± 0%      ~     (all equal)
RepoBranches_Decode-12         134k ± 0%      134k ± 0%      ~     (p=0.151 n=5+5)
RepoBranchesIDs_Encode-12      89.0 ± 0%      98.0 ± 0%   +10.11%  (p=0.008 n=5+5)
RepoBranchesIDs_Decode-12       443 ± 0%       443 ± 0%      ~     (all equal)
RepoSet_Encode-12             11.0M ± 0%     11.0M ± 0%      ~     (all equal)
RepoSet_Decode-12             5.52M ± 0%     5.52M ± 0%      ~     (p=0.548 n=5+5)
RepoSetIDs_Encode-12           87.0 ± 0%      87.0 ± 0%      ~     (all equal)
RepoSetIDs_Decode-12            449 ± 0%       449 ± 0%      ~     (all equal)
```